Add basic feature filtering

matteblair opened the pull request:
Let's consider this a first pass. There's a lot more to do here, both to refine the classes that are here and to add an interface for creating filters in the Unity editor, but there's enough code here that I wanted to start merging it back to master.

`PropertyRangeMatcher` currently throws an exception when the property in the tile data isn't the same type as the range limits (which is very frequently), so I don't recommend using it yet.

karimnaaji commented on Assets/Mapzen/VectorData/Filters/PropertyValueFeatureMatcher.cs:
```diff
@@ -0,0 +1,14 @@
+
+        protected override bool MatchesProperty(object property)
+        {
+            return ValueSet.Contains(property);
```
Using direct comparison here without type casting can lead to some issues, here is an example:

Trying to filter out all features that have a `height` value of `417`:
```c#
var buildingExtrusionFilter = new FeatureFilter().TakeAllFromCollections("buildings").Where(FeatureMatcher.HasPropertyWithValue("height", 417));
```

```
> ValueSet
Count=1
	[1]: 417
	Raw View: 
> ValueSet.Contains(property)
false
> property == 417
false
> (int)property
417
> (int)property == 417
true
```

A possibility is to type check using introspection first, if they can be represented as the same type, typecast to the type of the `ValueSet`. only then we can safely send it to the helper method `Contains`.

matteblair replied:
Yeah totally, I've realized we're going to need a stricter representation for feature properties than just `object`. I imagine this could use a "variant" type which is either a string, a boolean, or a double. 

This is kind of tricky to solve now, because `ValueSet` may not even contain objects of a single type!

matteblair commented on Assets/MeshData.cs:
```diff
@@ -4,25 +4,38 @@
+        foreach (var index in indices)
         {
-            indices.Add(indexOffset + meshData.indices[i]);
+            submesh.Indices.Add(index + offset);
```
Nice! Linq is so slick.

matteblair commented:
@karimnaaji I've addressed the issue you discovered, though I imagine that our approach here might change again.

I changed the implementation of `TryGetProperty` because in the previous implementation, if you call it with a type that doesn't match the property object, the cast produces a runtime exception and that worker thread just stops. I basically just moved the type checking responsibility to the calling code. I'm open to further suggestions on this too.

karimnaaji commented:
The new changes looks good to me, good to go once rebased.
